src/main/java/org/elasticsearch/script/expression/FieldMethodValueSource.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.script.expression;
+
+import java.io.IOException;
+import java.util.Map;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.queries.function.FunctionValues;
+import org.apache.lucene.queries.function.ValueSource;
+
+import org.elasticsearch.index.fielddata.AtomicFieldData;
+import org.elasticsearch.index.fielddata.AtomicNumericFieldData;
+import org.elasticsearch.index.fielddata.IndexFieldData;
+
+class FieldMethodValueSource extends ValueSource {
+
+    private final IndexFieldData<?> indexFieldData;
+    private final FieldMethodFactory.Method method;
+
+    FieldMethodValueSource(IndexFieldData<?> indexFieldData, String method)
+    {
+        this.indexFieldData = indexFieldData;
+
+        try {
+            this.method = FieldMethodFactory.Method.valueOf(method);
+        } catch (IllegalArgumentException illegalArgumentException) {
+            throw new IllegalArgumentException("Member method [" + method + "] does not exist.", illegalArgumentException);
+        }
+    }
+
+    @Override
+    public FunctionValues getValues(Map context, LeafReaderContext leaf) throws IOException {
+        AtomicFieldData leafData = indexFieldData.load(leaf);
+        assert(leafData instanceof AtomicNumericFieldData);
+
+        return method.getFactory().createFieldMethodFunctionValues(this, (AtomicNumericFieldData)leafData);
+    }
+
+    @Override
+    public boolean equals(Object o) {
```

Comment:
Same here as before, `indexFieldData` and `method` cannot be null. Add assertions to the ctor and simplify equals/hashCode?